### ISSUE TITLE
Android: Editing existing entry with history and saving causes invalid chars in protected fields

### ISSUE THREAD

**Issue by ThomasSHA** [cternes/openkeepass#48]
I'm currently playing around with your lib.
Opening, listing etc. works great.
But when i try to edit an entry and it contains protected fields (e.g. password), upon save, the password is "unreadable" for other keepass viewers. The testapp itself is capable of decoding the protected fields.

```
File extStore = Environment.getExternalStorageDirectory();
        mKeePassDatabaseFile = extStore.getAbsolutePath() + "/test_simple.kdbx";
        mDatabase = KeePassDatabase.getInstance(mKeePassDatabaseFile).openDatabase("test");
        Group parentGroup = mDatabase.getRoot().getGroups().get(0);
        Entry entry = mDatabase.getEntries().get(0);
        Entry newEntry = new EntryBuilder(entry).username("hello1").buildWithHistory();
        new GroupBuilder(parentGroup).removeEntry(entry).addEntry(newEntry).build();

        KeePassDatabase.write(mDatabase, "test", mKeePassDatabaseFile);

```
Before editing the entry was: username: "user1222", password: "pwd1"
The result entry is: username "hello1", but password is "c:w�" 

These results i get when:
- Entry has no history but i use buildWithHistory()
- Entry has history and I use buildWithHistory()
- Entry has history and I user build()

in short: only no history and with "build()" works...

Can you please tell me what I'm doing wrong? 😢 

Thanks in advance

**Comment by cternes** [comment]
I'll take a look at it.

**Comment by cternes** [comment]
If I test with the code you provided it works perfectly on my system when I open the database with KeePass.

    KeePassFile mDatabase = KeePassDatabase.getInstance(ResourceUtils.getResource("test1.kdbx")).openDatabase("test");
        Group parentGroup = mDatabase.getRoot().getGroups().get(0);
        Entry entry = mDatabase.getEntries().get(0);
        Entry newEntry = new EntryBuilder(entry).username("hello1").buildWithHistory();
        new GroupBuilder(parentGroup).removeEntry(entry).addEntry(newEntry).build();

        KeePassDatabase.write(mDatabase, "test", ResourceUtils.getResource("test1.kdbx"));

Can you provide a bit more details?

1. Is it working if you're trying it on a desktop system and not Android?
2. You mentioned that 

> The password is "unreadable" for other keepass viewers

what are these other keepass viewers? Does it work with the original KeePass program?
3. Can you provide a testcase that is not working? As mentioned, the code above works for me.

**Comment by ThomasSHA** [comment]
Well it seems to be a problem with Android only.
Tried it on the Desktop and it works without problems.
On Android the same code results in a "destroyed" password field.

to answer your questions:
1. It works on Desktop system 
2. I tried the "original" KeePass Password Safe v2.35 on Desktop. Keepass2Android and KeepassDroid
3. I tried the same Code, on Android and Desktop. Works on Desktop, doesn't work on Android...

Below the file in it's different stages...
Android after edit: https://dropfile.to/O4UQ6wV
Desktop after edit: https://dropfile.to/k7bxX2T
Befoer edit: https://dropfile.to/ZycWp2c

**Comment by ThomasSHA** [comment]
Just a short update:
I might have found a new "hint".
It seems like the gzip compression breaks the file on Android.

I just created a new kdbx file with the windows client and set compression to "none".
Now editing and saving files with you library seems to work without problems.

**Comment by cternes** [comment]
Thanks for the hint, I can reproduce the issue and trying to fix it.

**Comment by cternes** [comment]
I finally found the issue. The problem is not the gzip compression, it's the behavior of Simple XML on Android. It does not preserve the order of XML tags on Android. 

However, I managed to fix the issue.
